File: android/SalmonTrollingAssistant/app/src/main/java/com/example/salmontrollingassistant/presentation/viewmodels/AuthViewModel.kt
package com.example.salmontrollingassistant.presentation.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.salmontrollingassistant.domain.model.AuthCredentials
import com.example.salmontrollingassistant.domain.model.AuthResult
import com.example.salmontrollingassistant.domain.model.UserProfile
import com.example.salmontrollingassistant.domain.service.AuthenticationService
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class AuthViewModel @Inject constructor(
    private val authService: AuthenticationService
) : ViewModel() {
    
    private val _authState = MutableStateFlow<AuthState>(AuthState.Initial)
    val authState: StateFlow<AuthState> = _authState.asStateFlow()
    
    private val _currentUser = MutableStateFlow<UserProfile?>(null)
    val currentUser: StateFlow<UserProfile?> = _currentUser.asStateFlow()
    
    private val _isSignedIn = MutableStateFlow(false)
    val isSignedIn: StateFlow<Boolean> = _isSignedIn.asStateFlow()
    
    private val _isAnonymous = MutableStateFlow(false)
    val isAnonymous: StateFlow<Boolean> = _isAnonymous.asStateFlow()
    
    init {
        viewModelScope.launch {
            authService.getCurrentUser().collectLatest {
                _currentUser.value = it
            }
        }
        
        viewModelScope.launch {
            authService.isSignedIn().collectLatest {
                _isSignedIn.value = it
            }
        }
        
        viewModelScope.launch {
            authService.isAnonymousUser().collectLatest {
                _isAnonymous.value = it
            }
        }
    }
    
    fun createAccount(email: String, password: String, name: String?) {
        if (!validateCredentials(email, password)) {
            return
        }
        
        viewModelScope.launch {
            _authState.value = AuthState.Loading
            val result = authService.createAccount(
                AuthCredentials(email, password),
                name
            )
            _authState.value = if (result.success) {
                AuthState.Success(result)
            } else {
                AuthState.Error(result.error ?: "Unknown error")
            }
        }
    }
    
    fun signIn(email: String, password: String) {
        if (!validateCredentials(email, password)) {
            return
        }
        
        viewModelScope.launch {
            _authState.value = AuthState.Loading
            val result = authService.signIn(AuthCredentials(email, password))
            _authState.value = if (result.success) {
                AuthState.Success(result)
            } else {
                AuthState.Error(result.error ?: "Unknown error")
            }
        }
    }
    
    fun signOut() {
        viewModelScope.launch {
            _authState.value = AuthState.Loading
            val success = authService.signOut()
            _authState.value = if (success) {
                AuthState.Initial
            } else {
                AuthState.Error("Failed to sign out")
            }
        }
    }
    
    fun signInAnonymously() {
        viewModelScope.launch {
            _authState.value = AuthState.Loading
            val result = authService.signInAnonymously()
            _authState.value = if (result.success) {
                AuthState.Success(result)
            } else {
                AuthState.Error(result.error ?: "Unknown error")
            }
        }
    }
    
    fun convertAnonymousAccount(email: String, password: String, name: String?) {
        if (!validateCredentials(email, password)) {
            return
        }
        
        viewModelScope.launch {
            _authState.value = AuthState.Loading
            val result = authService.convertAnonymousAccount(
                AuthCredentials(email, password),
                name
            )
            _authState.value = if (result.success) {
                AuthState.Success(result)
            } else {
                AuthState.Error(result.error ?: "Unknown error")
            }
        }
    }
    
    fun resetAuthState() {
        _authState.value = AuthState.Initial
    }
    
    private fun validateCredentials(email: String, password: String): Boolean {
        if (email.isBlank() || !android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
            _authState.value = AuthState.Error("Invalid email address")
            return false
        }
        
        if (password.length < 6) {
            _authState.value = AuthState.Error("Password must be at least 6 characters")
            return false
        }
        
        return true
    }
    
    sealed class AuthState {
        object Initial : AuthState()
        object Loading : AuthState()
        data class Success(val result: AuthResult) : AuthState()
        data class Error(val message: String) : AuthState()
    }
}

File: SalmonTrollingAssistant/Presentation/ViewModels/AuthViewModel.swift
import Foundation
import Combine

class AuthViewModel: ObservableObject {
    private let authService: AuthenticationService
    private var cancellables = Set<AnyCancellable>()
    
    @Published var authState: AuthState = .initial
    @Published var currentUser: UserProfile?
    @Published var isSignedIn: Bool = false
    @Published var isAnonymous: Bool = false
    
    init(authService: AuthenticationService) {
        self.authService = authService
        
        // Subscribe to user changes
        authService.currentUser
            .assign(to: \.currentUser, on: self)
            .store(in: &cancellables)
        
        authService.isSignedIn
            .assign(to: \.isSignedIn, on: self)
            .store(in: &cancellables)
        
        authService.isAnonymousUser
            .assign(to: \.isAnonymous, on: self)
            .store(in: &cancellables)
    }
    
    func createAccount(email: String, password: String, name: String?) {
        guard validateCredentials(email: email, password: password) else {
            return
        }
        
        authState = .loading
        
        Task {
            let result = await authService.createAccount(
                credentials: AuthCredentials(email: email, password: password),
                name: name
            )
            
            await MainActor.run {
                if result.success {
                    authState = .success(result)
                } else {
                    authState = .error(result.error ?? "Unknown error")
                }
            }
        }
    }
    
    func signIn(email: String, password: String) {
        guard validateCredentials(email: email, password: password) else {
            return
        }
        
        authState = .loading
        
        Task {
            let result = await authService.signIn(
                credentials: AuthCredentials(email: email, password: password)
            )
            
            await MainActor.run {
                if result.success {
                    authState = .success(result)
                } else {
                    authState = .error(result.error ?? "Unknown error")
                }
            }
        }
    }
    
    func signOut() {
        authState = .loading
        
        Task {
            let success = await authService.signOut()
            
            await MainActor.run {
                if success {
                    authState = .initial
                } else {
                    authState = .error("Failed to sign out")
                }
            }
        }
    }
    
    func signInAnonymously() {
        authState = .loading
        
        Task {
            let result = await authService.signInAnonymously()
            
            await MainActor.run {
                if result.success {
                    authState = .success(result)
                } else {
                    authState = .error(result.error ?? "Unknown error")
                }
            }
        }
    }
    
    func convertAnonymousAccount(email: String, password: String, name: String?) {
        guard validateCredentials(email: email, password: password) else {
            return
        }
        
        authState = .loading
        
        Task {
            let result = await authService.convertAnonymousAccount(
                credentials: AuthCredentials(email: email, password: password),
                name: name
            )
            
            await MainActor.run {
                if result.success {
                    authState = .success(result)
                } else {
                    authState = .error(result.error ?? "Unknown error")
                }
            }
        }
    }
    
    func resetAuthState() {
        authState = .initial
    }
    
    private func validateCredentials(email: String, password: String) -> Bool {
        let emailRegex = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,64}"
        let emailPredicate = NSPredicate(format: "SELF MATCHES %@", emailRegex)
        
        guard !email.isEmpty, emailPredicate.evaluate(with: email) else {
            authState = .error("Invalid email address")
            return false
        }
        
        guard password.count >= 6 else {
            authState = .error("Password must be at least 6 characters")
            return false
        }
        
        return true
    }
    
    enum AuthState {
        case initial
        case loading
        case success(AuthResult)
        case error(String)
    }
}